service/history/decision/handler_test.go

```diff
@@ -80,34 +109,186 @@ func (s *DecisionHandlerSuite) TearDownTest() {
+					Times(1).
+					Return(nil, &persistence.TimeoutError{Msg: "failed to get start event: request timeout"})
+				shardContext.EXPECT().GetShardID().Return(_testShardID).Times(1)
+			},
+			expectErr: true,
+		},
+		{
+			name:     "test HandleDecisionTaskScheduled - first decision task scheduled failure",
+			domainID: _testDomainUUID,
+			mutablestate: &persistence.WorkflowMutableState{
+				ExecutionInfo: &persistence.WorkflowExecutionInfo{
+					DecisionScheduleID: -23,
+					LastProcessedEvent: -23,
+				},
+				BufferedEvents: append([]*types.HistoryEvent{}, &types.HistoryEvent{}),
+			},
+			expectCalls: func(shardContext *shard.MockContext) {
+				eventsCache := events.NewMockCache(s.controller)
+				shardContext.EXPECT().GetEventsCache().Times(1).Return(eventsCache)
+				eventsCache.EXPECT().
+					GetEvent(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+					Times(1).
+					Return(&types.HistoryEvent{}, nil)
+				shardContext.EXPECT().GetShardID().Return(_testShardID).Times(1)
+				shardContext.EXPECT().GenerateTransferTaskIDs(gomock.Any()).Times(1).Return([]int64{}, errors.New("some random error to avoid going too deep in call stack unrelated to this unit"))
+			},
+			expectErr:       true,
+			isfirstDecision: true,
+		},
+		{
+			name:     "test HandleDecisionTaskScheduled - first decision task scheduled success",
+			domainID: _testDomainUUID,
+			mutablestate: &persistence.WorkflowMutableState{
+				ExecutionInfo: &persistence.WorkflowExecutionInfo{
+					DecisionScheduleID: -23,
+					LastProcessedEvent: -23,
+				},
+			},
+			expectCalls: func(shardContext *shard.MockContext) {
+				eventsCache := events.NewMockCache(s.controller)
+				shardContext.EXPECT().GetEventsCache().Times(1).Return(eventsCache)
+				eventsCache.EXPECT().
+					GetEvent(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+					Times(1).
+					Return(&types.HistoryEvent{}, nil)
+				shardContext.EXPECT().GetShardID().Return(_testShardID).Times(1)
+				shardContext.EXPECT().GenerateTransferTaskIDs(gomock.Any()).Times(1).Return([]int64{}, errors.New("some random error to avoid going too deep in call stack unrelated to this unit"))
+			},
+			expectErr:       true,
+			isfirstDecision: true,
+		},
+	}
+	for _, test := range tests {
+		s.Run(test.name, func() {
+			request := &types.ScheduleDecisionTaskRequest{
+				DomainUUID: test.domainID,
+				WorkflowExecution: &types.WorkflowExecution{
+					WorkflowID: _testWorkflowID,
+					RunID:      _testRunID,
+				},
+				IsFirstDecision: test.isfirstDecision,
+			}
+			shardContext := shard.NewMockContext(s.controller)
+			test.expectCalls(shardContext)
+			s.expectHandleDecisionTaskScheduledCalls(test.domainID, test.mutablestate, shardContext)
+
+			handler := *s.decisionHandler
```

Comment:
nit: no need to dereference the pointer. you can also get rid of local `handler` variable all together